File: frontend/erp-portfolio-web/src/types/pedido.ts
/**
 * =====================================================================
 * Arquivo....: pedido.ts
 * Versão.....: 1.2.0
 * Data.......: 23/09/2026
 * Descrição..: Tipos do módulo de Pedidos, espelhando os DTOs da API
 *              (PedidoRespostaDto, PedidoResumoDto, PedidoCriacaoDto,
 *              PedidoItemEntradaDto e PedidoFiltroDto). Status e forma de
 *              pagamento chegam e saem como TEXTO.
 * ---------------------------------------------------------------------
 * Histórico de alterações:
 *   1.0.0 - 21/09/2026 - Criação do arquivo.
 *   1.1.0 - 23/09/2026 - Vendedor e % de comissão congelada (etapa 10).
 *   1.2.0 - 24/09/2026 - quantidadeDevolvida por item e valorDevolvido (etapa 14).
 * =====================================================================
 */

export type StatusPedido = 'Rascunho' | 'Confirmado' | 'Cancelado'

export type FormaPagamento = 'Dinheiro' | 'Pix' | 'Boleto' | 'Cartao'

export const OPCOES_FORMA_PAGAMENTO: { value: FormaPagamento; label: string }[] = [
  { value: 'Dinheiro', label: 'Dinheiro' },
  { value: 'Pix', label: 'Pix' },
  { value: 'Boleto', label: 'Boleto' },
  { value: 'Cartao', label: 'Cartão' },
]

export const OPCOES_STATUS_PEDIDO: { value: StatusPedido; label: string }[] = [
  { value: 'Rascunho', label: 'Rascunho' },
  { value: 'Confirmado', label: 'Confirmado' },
  { value: 'Cancelado', label: 'Cancelado' },
]

export interface PedidoItem {
  id: number
  produtoId: number
  produtoNome: string
  sku: string
  unidade: string
  quantidade: number
  /** Preço do produto no momento em que o item foi adicionado (congelado). */
  precoUnitario: number
  descontoPercentual: number
  subtotal: number
  /** Quanto deste item já voltou em devoluções (0 no orçamento e em pedido sem devolução). */
  quantidadeDevolvida: number
}

export interface Pedido {
  /** Também é o número do pedido. */
  id: number
  clienteId: number
  clienteNome: string
  clienteDocumento: string
  /** Data/hora ISO 8601 em UTC. */
  dataPedido: string
  status: StatusPedido
  formaPagamento: FormaPagamento | null
  vendedorId: number | null
  vendedorNome: string | null
  /** % de comissão do vendedor congelada ao confirmar; nula no rascunho e em pedidos antigos. */
  percentualComissao: number | null
  descontoPercentual: number
  subtotalItens: number
  valorTotal: number
  itens: PedidoItem[]
  /** Soma das devoluções do pedido (etapa 14). */
  valorDevolvido: number
}

/** Uma linha da listagem (sem os itens). */
export interface PedidoResumo {
  id: number
  clienteId: number
  clienteNome: string
  dataPedido: string
  status: StatusPedido
  valorTotal: number
  quantidadeItens: number
}

/** Item enviado à API: sem preço (o servidor copia do produto). */
export interface PedidoItemEntrada {
  produtoId: number
  quantidade: number
  descontoPercentual: number
}

/** Corpo do POST e do PUT (o PUT substitui o rascunho inteiro). */
export interface PedidoEntrada {
  clienteId: number
  formaPagamento: FormaPagamento | null
  /** Opcional no rascunho; obrigatório para confirmar. */
  vendedorId: number | null
  descontoPercentual: number
  itens: PedidoItemEntrada[]
}

/** Corpo do PATCH /confirmar: número de parcelas e intervalo entre elas, para as contas a receber. */
export interface PedidoConfirmarEntrada {
  numeroParcelas: number
  intervaloDias: number
}

export interface PedidoFiltro {
  /** Número do pedido (ex.: 12 ou #12) ou trecho do nome do cliente. */
  busca?: string
  status?: StatusPedido
  pagina: number
  tamanhoPagina: number
}

File: backend/ErpPortfolio.Api/Models/Pedido.cs
// =====================================================================================
// Arquivo....: Pedido.cs
// Versão.....: 1.1.0
// Data.......: 21/09/2026
// Descrição..: Entidade de domínio que representa um pedido de venda: cabeçalho com cliente,
//              status, forma de pagamento, desconto do pedido e o total calculado.
// -------------------------------------------------------------------------------------
// Banco......: PostgreSQL - erp_portfolio_db
// Tabelas....: public.pedidos (FK cliente_id -> public.clientes)
// Fontes.....: Mapeada em ErpPortfolioDbContext.Pedidos (EF Core / Npgsql).
//              Colunas: id, cliente_id, data_pedido, status, forma_pagamento,
//              desconto_percentual, valor_total (o mapeamento de nomes fica no DbContext).
//              O total é recalculado pelo serviço a cada gravação (Services/CalculoPedido.cs).
// -------------------------------------------------------------------------------------
// Histórico de alterações:
//   1.0.0 - 21/09/2026 - Criação do arquivo.
//   1.1.0 - 23/09/2026 - Vendedor e % de comissão congelada ao confirmar (etapa 10).
// =====================================================================================

namespace ErpPortfolio.Api.Models;

public class Pedido
{
    /// <summary>Também é o número do pedido.</summary>
    public int Id { get; set; }

    public int ClienteId { get; set; }

    public Cliente? Cliente { get; set; }

    /// <summary>Data/hora de criação em UTC; não é editável.</summary>
    public DateTime DataPedido { get; set; }

    public StatusPedido Status { get; set; } = StatusPedido.Rascunho;

    /// <summary>Opcional no rascunho; obrigatória para confirmar.</summary>
    public FormaPagamento? FormaPagamento { get; set; }

    /// <summary>Opcional no rascunho; obrigatório (e ativo) para confirmar. Pedidos antigos ficam sem.</summary>
    public int? VendedorId { get; set; }

    public Vendedor? Vendedor { get; set; }

    /// <summary>% de comissão do vendedor copiada ao confirmar; nula no rascunho e nunca muda depois.</summary>
    public decimal? PercentualComissao { get; set; }

    /// <summary>Desconto em % sobre a soma dos itens (0 a 100).</summary>
    public decimal DescontoPercentual { get; set; }

    /// <summary>Total do pedido em reais, gravado a cada salvamento.</summary>
    public decimal ValorTotal { get; set; }

    public List<PedidoItem> Itens { get; set; } = [];
}
